proposals/0000-likelihood-annotations.rst

```diff
@@ -0,0 +1,264 @@
+Add Likelihood Annotations
+==============
+
+.. proposal-number:: _.
+.. trac-ticket:: _.
+.. implemented:: _.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/182>`_.
+.. sectnum::
+.. contents::
+
+A lot of performance can be gained by optimizing for hot code paths.
+However GHC currently does not allow users to actually give hints about which code
+paths are hot code paths.
+
+This proposal changes this.
+
+
+Motivation
+------------
+
+Programmers often know a good deal more about program behaviour than we can currently
+communicate to the compiler. This leads to missed optimization opportunities.
+
+Consider types like ``data Either a b = Left a | Right b``
+
+Left is often used to represent an exceptional case. Making Right the common case.
+
+Based on this information and given code like ``either f g x`` we might want
+to inline g but not f for example.
+
+The same logic applies to many other cases. For example Lists are rarely empty.
```

Comment:
I'm doubtful about the usefulness of annotating data types, especially if they're not for a very specific purpose. `Either` is not always used for exceptions, sometimes it's just used as a generic sum type. Likewise, there are many situations where lists can likely be empty, e.g. when passing a list of extra configuration options.

Reply:
It's not useful for every type. But for some types it's really useful.

For streams defined via Done/Yield/Step `Done` will always be very unlikely. Just is generally more likely than Nothing. `:` is more likely than `Nil` and so on.

The alternative to annotating types like these would be us expecting the user to to annotate every use of the type.

Either was probably not the best example. But I do think it would be good to attach a weak preference to `Right` in the either case nevertheless.